### ISSUE TITLE
inconsistent behaviour of InvalidCast

### ISSUE THREAD

**Issue by SignpostMarv** [vimeo/psalm#1398]
# issue
* array manually type-hinted to `scalar|array|object|resource|null` does not flag up `MixedAssignment` ***(could go either way on this one)***

## behaves as expected
* typehinted array correctly flags up `InvalidCast`
* array manually typed-hinted to `scalar|array|object|null` does flag up `InvalidCast`

## does not behave as expected
* array assigned to mixed does not flag up `InvalidCast`
* array manually type-hinted to `scalar|array|object|resource|null` (aka `mixed`) does not flag up `InvalidCast`

# repro
* https://psalm.dev/r/fa71008e9a
* using 119687f
```php
<?php
$a = [];

echo (string) $a;

/**
* @var mixed
*/
$b = [];

echo (string) $b;

/**
* @var scalar|array|object|resource|null
*/
$c = [];

echo (string) $c;

/**
* @var scalar|array|object|null
*/
$d = [];

echo (string) $d;
```
> Psalm output (using commit 119687f): 
> 
> ERROR: InvalidCast - 4:6 - array<empty, empty> cannot be cast to string
> 
> INFO: MixedAssignment - 9:1 - Cannot assign $b to a mixed type
> 
> ERROR: InvalidCast - 25:6 - array<array-key, mixed>|scalar|object|null cannot be cast to string

**Comment by muglug** [comment]
It's really just a failure of

```php
/**
* @var array|resource
*/
$c = [];

echo (string) $c;
```

to emit an error
